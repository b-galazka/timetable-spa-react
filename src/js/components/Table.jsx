import React, { Component } from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';
import propTypes from 'prop-types';

import detectIE10 from '../utils/detectIE10';
import LessonHour from './LessonHour';

import texts from '../../json/texts';

import '../../scss/table.scss';

function mapStateToProps(state) {

    const { timetableObject, hours } = state;

    return { timetableObject: timetableObject.data, hours };
}

class Table extends Component {

    constructor() {

        super();

        this.id = 0;
        this.tableCells = [];
    }

    render() {

        const numberOfLastHour = this.getNumberOfLastHour();

        return (
            <div className="table">
                <header className="table__header table__header--ordinal-numbers">
                    {texts.number}
                </header>

                {this.renderOridinalNumbers(numberOfLastHour)}

                <header className="table__header table__header--hours">
                    {texts.hour}
                </header>

                {this.renderHours(numberOfLastHour)}

                {this.renderDays(numberOfLastHour)}
            </div>
        );
    }

    componentDidMount() {

        this.applyPrefixes();
    }

    renderOridinalNumbers(number) {

        const cells = [];

        for (let i = 1; i <= number; i++) {

            cells.push(
                <div 
                    className="table__cell table__cell--ordinal-number" 
                    key={++this.id}
                    style={{ order: i + 1 }}
                    ref={(el) => { this.tableCells.push(el) }}
                >
                    {i}
                </div>
            );
        }

        return cells;    
    }

    renderHours(number) {

        const { hours } = this.props;
        const cells = [];

        for (let i = 0; i < number; i++) {

            const hour = hours[i];

            cells.push(
                <div 
                    className="table__cell table__cell--hour" 
                    key={hour._id}
                    style={{ order: i + 2 }}
                    data-number={i + 1}
                    ref={(el) => { this.tableCells.push(el) }}
                >
                    {hour.start} - {hour.end}
                </div>
            );
        }

        return cells;
    }

    renderDays(lastHour) {

        const { days } = texts;

        return days.map((day, index) => {

            return [
                <header
                    className={

                        classNames({
                            'table__header': true,
                            'table__header--mobile-hidden': !this.areLessonsInDay(index)
                        })
                    }

                    key={++this.id}
                >
                    {day}
                </header>,

                this.renderLessons(index, lastHour)
            ];
        });
    }

    renderLessons(dayNumber, lastHour) {

        const { timetable, type } = this.props.timetableObject;
        const day = timetable[dayNumber];

        let cells = [];

        if (day.length === 0) {

            this.pushEmptyCells(cells, lastHour);
        } else {

            cells = day.map((hour, index) => (
                <LessonHour
                    lessons={hour}
                    timetableObjectType={type}
                    order={index + 2}
                    number={index + 1}
                    key={++this.id}
                />
            ));

            this.pushEmptyCells(cells, lastHour);
        }

        return cells;
    }

    pushEmptyCells(cells, lastHour) {

        for (let i = cells.length; i < lastHour; i++) {

            cells.push(
                <LessonHour
                    mobileHidden
                    order={i + 2}
                    key={++this.id}
                />
            );
        }
    }

    getNumberOfLastHour() {

        const { timetable } = this.props.timetableObject;

        return timetable.reduce((prev, arr) => (prev > arr.length) ? prev : arr.length, 0);
    }

    areLessonsInDay(dayNumber) {

        const { timetable } = this.props.timetableObject;

        return timetable[dayNumber].some(
            lessonsArr => lessonsArr[0] && lessonsArr.length > 0
        );
    }

    applyPrefixes() {

        if (detectIE10()) {

            this.tableCells.forEach((cell) => {

                const { order } = cell.style;

                cell.style.msFlexOrder = order;
            });
        }
    }
}

Table.propTypes = {

    // redux
    hours: propTypes.arrayOf(propTypes.shape({
        start: propTypes.string.isRequired,
        end: propTypes.string.isRequired,
        _id: propTypes.string.isRequired
    })).isRequired,

    timetableObject: propTypes.shape({
        timetable: propTypes.arrayOf(
            propTypes.arrayOf(
                propTypes.arrayOf(
                    propTypes.shape({
                        _id: propTypes.string.isRequired,
                        subject: propTypes.string.isRequired,
                        class: propTypes.string,
                        classroom: propTypes.string,
                        teacherName: propTypes.string,
                        teacherSlug: propTypes.string
                    })
                )
            ).isRequired
        ).isRequired,

        type: propTypes.oneOf(['teacher', 'class', 'classroom']).isRequired
    }).isRequired,
};

export default connect(mapStateToProps)(Table);