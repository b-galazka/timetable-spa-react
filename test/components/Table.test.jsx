import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';

import Table from '../../src/js/components/Table';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../src/js/components/LessonHour', () => 'LessonHour');

jest.mock(
    '../../src/json/texts',
    () => ({
        number: 'number',
        hour: 'hour',
        days: ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday']
    })
);

describe('Table component', () => {

    const { userAgent } = navigator;

    const hours = [
        { start: '1:00', end: '2:00', _id: 'hour1' },
        { start: '2:00', end: '3:00', _id: 'hour2' },
        { start: '3:00', end: '4:00', _id: 'hour3' },
        { start: '4:00', end: '5:00', _id: 'hour4' },
        { start: '5:00', end: '6:00', _id: 'hour5' }
    ];

    const teacherTimetableObject = {
        timetable: [
            [
                [
                    {
                        subject: 'math',
                        class: '3ET',
                        classroom: '200',
                        _id: 'lesson1'
                    },
                    {
                        subject: 'math',
                        class: '4ET',
                        classroom: '200',
                        _id: 'lesson2'
                    }
                ]
            ],
            [],
            [
                [null],
                [
                    {
                        subject: 'math',
                        class: '3ET',
                        classroom: '200',
                        _id: 'lesson1'
                    }
                ],
                [
                    {
                        subject: 'math',
                        class: '4ET',
                        classroom: '200',
                        _id: 'lesson2'
                    }
                ],
                [
                    {
                        subject: 'math',
                        class: '4ET',
                        classroom: '200',
                        _id: 'lesson2'
                    }
                ]
            ],
            [],
            []
        ],

        type: 'teacher'
    };

    it('should render teacher\'s timetable', () => {

        const tree = renderer.create(
            <Table.WrappedComponent
                hours={hours}
                timetableObject={teacherTimetableObject}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render class\'s timetable', () => {

        const tree = renderer.create(
            <Table.WrappedComponent

                hours={hours}

                timetableObject={{
                    timetable: [
                        [
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'AB',
                                    teacherName: 'AB name',
                                    classroom: '200',
                                    _id: 'lesson1'
                                }
                            ],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'AB',
                                    teacherName: 'AB name',
                                    classroom: '200',
                                    _id: 'lesson2'
                                }
                            ]
                        ],
                        [],
                        [
                            [null],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'AB',
                                    teacherName: null,
                                    classroom: '200',
                                    _id: 'lesson1'
                                }
                            ],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'XY',
                                    teacherName: 'XY name',
                                    classroom: '200',
                                    _id: 'lesson2'
                                }
                            ],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'XYZ',
                                    teacherName: null,
                                    classroom: '200',
                                    _id: 'lesson2'
                                }
                            ]
                        ],
                        [],
                        []
                    ],

                    type: 'class'
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render classroom\'s timetable', () => {

        const tree = renderer.create(
            <Table.WrappedComponent

                hours={hours}

                timetableObject={{
                    timetable: [
                        [
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'AB',
                                    teacherName: 'AB name',
                                    class: '3ET',
                                    _id: 'lesson1'
                                }
                            ],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'AB',
                                    teacherName: 'AB name',
                                    class: '4ET',
                                    _id: 'lesson2'
                                }
                            ]
                        ],
                        [],
                        [
                            [null],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'AB',
                                    teacherName: null,
                                    class: '1AL',
                                    _id: 'lesson1'
                                }
                            ],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'XY',
                                    teacherName: 'XY name',
                                    class: '3BL',
                                    _id: 'lesson2'
                                }
                            ],
                            [
                                {
                                    subject: 'math',
                                    teacherSlug: 'XYZ',
                                    teacherName: null,
                                    class: '3CT',
                                    _id: 'lesson2'
                                }
                            ]
                        ],
                        [],
                        []
                    ],

                    type: 'classroom'
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should add style.msFlexOrder to all table cells ' +
        'if app is running on IE10', () => {

        Object.defineProperty(navigator, 'userAgent', {
            value: 'lorem MSIE ipsum',
            configurable: true
        });

        const table = mount(
            <Table.WrappedComponent
                hours={hours}
                timetableObject={teacherTimetableObject}
            />
        );

        const everyCellHasPrefix = Array.prototype.every.call(
            table.find('.table__cell'),
            (elem) => {

                const elemStyles = elem.getDOMNode().style;

                return (elemStyles.order === elemStyles.msFlexOrder);
            }
        );

        expect(everyCellHasPrefix).toBe(true);
    });

    it('should add style.msFlexOrder to all table cells ' +
        'if app is running on IE10 (regex)', () => {

        Object.defineProperty(navigator, 'userAgent', {
            value: 'lorem Tridentxrv:11. ipsum',
            configurable: true
        });

        const table = mount(
            <Table.WrappedComponent
                hours={hours}
                timetableObject={teacherTimetableObject}
            />
        );

        const everyCellHasPrefix = Array.prototype.every.call(
            table.find('.table__cell'),
            (elem) => {

                const elemStyles = elem.getDOMNode().style;

                return (elemStyles.order === elemStyles.msFlexOrder);
            }
        );

        expect(everyCellHasPrefix).toBe(true);
    });

    afterEach(() => {

        Object.defineProperty(navigator, 'userAgent', {
            value: userAgent,
            configurable: true
        });
    });
});