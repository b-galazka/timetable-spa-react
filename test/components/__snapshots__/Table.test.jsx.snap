// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should render class's timetable 1`] = `
<div
  className="table"
>
  <header
    className="table__header table__header--ordinal-numbers"
  >
    number
  </header>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 2,
      }
    }
  >
    1
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 3,
      }
    }
  >
    2
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 4,
      }
    }
  >
    3
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 5,
      }
    }
  >
    4
  </div>
  <header
    className="table__header table__header--hours"
  >
    hour
  </header>
  <div
    className="table__cell table__cell--hour"
    data-number={1}
    style={
      Object {
        "order": 2,
      }
    }
  >
    1:00
     - 
    2:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={2}
    style={
      Object {
        "order": 3,
      }
    }
  >
    2:00
     - 
    3:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={3}
    style={
      Object {
        "order": 4,
      }
    }
  >
    3:00
     - 
    4:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={4}
    style={
      Object {
        "order": 5,
      }
    }
  >
    4:00
     - 
    5:00
  </div>
  <header
    className="table__header"
  >
    Monday
  </header>
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson1",
          "classroom": "200",
          "subject": "math",
          "teacherName": "AB name",
          "teacherSlug": "AB",
        },
      ]
    }
    number={1}
    order={2}
    timetableObjectType="class"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "classroom": "200",
          "subject": "math",
          "teacherName": "AB name",
          "teacherSlug": "AB",
        },
      ]
    }
    number={2}
    order={3}
    timetableObjectType="class"
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Tuesday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header"
  >
    Wednesday
  </header>
  <LessonHour
    lessons={
      Array [
        null,
      ]
    }
    number={1}
    order={2}
    timetableObjectType="class"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson1",
          "classroom": "200",
          "subject": "math",
          "teacherName": null,
          "teacherSlug": "AB",
        },
      ]
    }
    number={2}
    order={3}
    timetableObjectType="class"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "classroom": "200",
          "subject": "math",
          "teacherName": "XY name",
          "teacherSlug": "XY",
        },
      ]
    }
    number={3}
    order={4}
    timetableObjectType="class"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "classroom": "200",
          "subject": "math",
          "teacherName": null,
          "teacherSlug": "XYZ",
        },
      ]
    }
    number={4}
    order={5}
    timetableObjectType="class"
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Thursday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Friday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
</div>
`;

exports[`Table component should render classroom's timetable 1`] = `
<div
  className="table"
>
  <header
    className="table__header table__header--ordinal-numbers"
  >
    number
  </header>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 2,
      }
    }
  >
    1
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 3,
      }
    }
  >
    2
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 4,
      }
    }
  >
    3
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 5,
      }
    }
  >
    4
  </div>
  <header
    className="table__header table__header--hours"
  >
    hour
  </header>
  <div
    className="table__cell table__cell--hour"
    data-number={1}
    style={
      Object {
        "order": 2,
      }
    }
  >
    1:00
     - 
    2:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={2}
    style={
      Object {
        "order": 3,
      }
    }
  >
    2:00
     - 
    3:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={3}
    style={
      Object {
        "order": 4,
      }
    }
  >
    3:00
     - 
    4:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={4}
    style={
      Object {
        "order": 5,
      }
    }
  >
    4:00
     - 
    5:00
  </div>
  <header
    className="table__header"
  >
    Monday
  </header>
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson1",
          "class": "3ET",
          "subject": "math",
          "teacherName": "AB name",
          "teacherSlug": "AB",
        },
      ]
    }
    number={1}
    order={2}
    timetableObjectType="classroom"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "class": "4ET",
          "subject": "math",
          "teacherName": "AB name",
          "teacherSlug": "AB",
        },
      ]
    }
    number={2}
    order={3}
    timetableObjectType="classroom"
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Tuesday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header"
  >
    Wednesday
  </header>
  <LessonHour
    lessons={
      Array [
        null,
      ]
    }
    number={1}
    order={2}
    timetableObjectType="classroom"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson1",
          "class": "1AL",
          "subject": "math",
          "teacherName": null,
          "teacherSlug": "AB",
        },
      ]
    }
    number={2}
    order={3}
    timetableObjectType="classroom"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "class": "3BL",
          "subject": "math",
          "teacherName": "XY name",
          "teacherSlug": "XY",
        },
      ]
    }
    number={3}
    order={4}
    timetableObjectType="classroom"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "class": "3CT",
          "subject": "math",
          "teacherName": null,
          "teacherSlug": "XYZ",
        },
      ]
    }
    number={4}
    order={5}
    timetableObjectType="classroom"
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Thursday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Friday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
</div>
`;

exports[`Table component should render teacher's timetable 1`] = `
<div
  className="table"
>
  <header
    className="table__header table__header--ordinal-numbers"
  >
    number
  </header>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 2,
      }
    }
  >
    1
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 3,
      }
    }
  >
    2
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 4,
      }
    }
  >
    3
  </div>
  <div
    className="table__cell table__cell--ordinal-number"
    style={
      Object {
        "order": 5,
      }
    }
  >
    4
  </div>
  <header
    className="table__header table__header--hours"
  >
    hour
  </header>
  <div
    className="table__cell table__cell--hour"
    data-number={1}
    style={
      Object {
        "order": 2,
      }
    }
  >
    1:00
     - 
    2:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={2}
    style={
      Object {
        "order": 3,
      }
    }
  >
    2:00
     - 
    3:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={3}
    style={
      Object {
        "order": 4,
      }
    }
  >
    3:00
     - 
    4:00
  </div>
  <div
    className="table__cell table__cell--hour"
    data-number={4}
    style={
      Object {
        "order": 5,
      }
    }
  >
    4:00
     - 
    5:00
  </div>
  <header
    className="table__header"
  >
    Monday
  </header>
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson1",
          "class": "3ET",
          "classroom": "200",
          "subject": "math",
        },
        Object {
          "_id": "lesson2",
          "class": "4ET",
          "classroom": "200",
          "subject": "math",
        },
      ]
    }
    number={1}
    order={2}
    timetableObjectType="teacher"
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Tuesday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header"
  >
    Wednesday
  </header>
  <LessonHour
    lessons={
      Array [
        null,
      ]
    }
    number={1}
    order={2}
    timetableObjectType="teacher"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson1",
          "class": "3ET",
          "classroom": "200",
          "subject": "math",
        },
      ]
    }
    number={2}
    order={3}
    timetableObjectType="teacher"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "class": "4ET",
          "classroom": "200",
          "subject": "math",
        },
      ]
    }
    number={3}
    order={4}
    timetableObjectType="teacher"
  />
  <LessonHour
    lessons={
      Array [
        Object {
          "_id": "lesson2",
          "class": "4ET",
          "classroom": "200",
          "subject": "math",
        },
      ]
    }
    number={4}
    order={5}
    timetableObjectType="teacher"
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Thursday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
  <header
    className="table__header table__header--mobile-hidden"
  >
    Friday
  </header>
  <LessonHour
    mobileHidden={true}
    order={2}
  />
  <LessonHour
    mobileHidden={true}
    order={3}
  />
  <LessonHour
    mobileHidden={true}
    order={4}
  />
  <LessonHour
    mobileHidden={true}
    order={5}
  />
</div>
`;
